distributed/rpc/parameter_server/rpc_param_server.py

```diff
@@ -0,0 +1,226 @@
+    def __init__(self):
+        super().__init__()
+        model = Net()
+        self.model = model
+
+    def forward(self, inp):
+        out = self.model(inp)
+        return out
+
+    # Use dist autograds to retrieve gradients accumulated for this model.
+    # Primarily used for verification.
+    def get_dist_gradients(self, cid):
+        grads = dist_autograd.get_gradients(cid)
+        return grads
+
+    # Wrap local parameters in a RRef. Needed for building the
+    # DistributedOptimizer which optimizes paramters remotely.
+    def get_param_rrefs(self):
+        param_rrefs = [rpc.RRef(param) for param in self.model.parameters()]
+        return param_rrefs
+
+
+param_server = None
+global_lock = Lock()
+# Ensure that we get only one handle to the ParameterServer.
+
+
+def get_parameter_server():
+    global param_server
+    with global_lock:
+        if not param_server:
+            # construct it once
+            param_server = ParameterServer()
+        return param_server
+
+
+def run_parameter_server(rank, world_size):
+    # The parameter server just acts as a host for the model and responds to
+    # requests from trainers, hence it does not need to run a loop.
+    # rpc.shutdown() will wait for all workers to complete by default, which
+    # in this case means that the parameter server will wait for all trainers
+    # to complete, and then exit.
+    rpc.init_rpc(name="parameter_server", rank=rank, world_size=world_size)
+    rpc.shutdown()
+
+
+# --------- Trainers --------------------
+
+# nn.Module corresponding to the network trained by this trainer. The
+# forward() method simply invokes the network on the given parameter
+# server.
+class TrainerNet(nn.Module):
+    def __init__(self):
+        super().__init__()
+        model = Net()
+        self.model = model
+        # TODO take this in as an arg
+        self.param_server_rref = rpc.remote(
+            "parameter_server", get_parameter_server, args=())
+
+    def get_global_param_rrefs(self):
+        remote_params = remote_method(
+            ParameterServer.get_param_rrefs,
+            self.param_server_rref)
+        return remote_params
+
+    def forward(self, x, cid):
+        model_output = remote_method(
+            ParameterServer.forward, self.param_server_rref, x)
+        return model_output
+
+
+def run_training_loop(rank, train_loader, test_loader):
+    # Runs the typical nueral network forward + backward + optimizer step, but
+    # in a distributed fashion.
+    net = TrainerNet()
+    for i, (data, target) in enumerate(train_loader):
+        with dist_autograd.context() as cid:
+            print("Training batch {}".format(i))
+            model_output = net(data, cid)
+            loss = F.nll_loss(model_output, target)
+            print(loss)
+            dist_autograd.backward([loss])
+            param_rrefs = net.get_global_param_rrefs()
+            opt = DistributedOptimizer(optim.SGD, param_rrefs, lr=0.03)
```

Comment:
Do we have to create an opt per iteration? Can this be done before the for loop?